### ISSUE TITLE
ArraySchemas proto format

### ISSUE THREAD

**Issue by fhackenberger** [protostuff/protostuff#139]
Hi!

I'm currently working on creating a .proto file for the RuntimeSchema in protostuff 1.3.6. I'm building on https://github.com/inwebby/protostuff-runtime-proto which was written for protostuff 1.0.9. Since then the way how arrays are mapped has changed. Reading ArraySchemas, I'm suspecting that it might be impossible to define a .proto file for the way how arrays are mapped, since the type of the actual array message fields itself depends on the id field.

```proto
 message ArrayScalar {
  optional uint32 id = 33;
  optional uint32 len = 1;
  repeated XXX data = 2;
}
```

My question is, how would I define the data field (marked XXX above) in .proto language, when the type of the data fields is only determined after reading and looking at the id field?

**Comment by fhackenberger** [comment]
@kshchepanovskyi It would be awesome if you could have a quick look on that. Being able to fix protostuff-runtime-proto opens up protostuff for a lot more usecases, as it enables people to use other languages than Java to interoperate. I'm using it for returning protobuf from a ReST service and using the result in Javascript.

**Comment by kshchepanovskyi** [comment]
> I'm suspecting that it might be impossible to define a .proto file for the way how arrays are mapped

In general, it's a conceptual problem, and there is no trivial solution. Maybe new [Any](https://developers.google.com/protocol-buffers/docs/proto3?hl=en#any) type (introduced in proto3) might help - I will check.

**Comment by fhackenberger** [comment]
Thanks for getting back to me! I looked at the Any time and I think that would be more of a workaround. How about changing the code to produce data matching a schema like this with the next release:

 message ArrayScalar {
  optional uint32 id = 33;
  optional uint32 len = 1;
  repeated uint32 dataInt = 2;
  repeated bool dataBool = 3;
  repeated double dataDouble = 4;
...
}

using the ID_XXX constants from RuntimeFieldFactory for the tags?

**Comment by kshchepanovskyi** [comment]
Need to check. In any case, it would be nice to have a formal model how `RuntimeSchema` is mapped to `proto` structure.

**Comment by kshchepanovskyi** [comment]
> How about changing the code to produce data matching a schema like this with the next release

We cannot change binary format, backward compatibility on binary level is required.

If we have to change binary format - then it will be done only in next major release (that will introduce many other changes - `2.0`).

**Comment by fhackenberger** [comment]
How about an option to enable it?

**Comment by kshchepanovskyi** [comment]
I think `id` is not needed in this case (anyway it is available only after serialization). As far as I understand, it is stored there by some historical reasons (polymorphic serialization). 

All what is needed is in the schema instance:

![image](https://cloud.githubusercontent.com/assets/4040120/10232414/503f7038-6891-11e5-9914-e3940e2966f1.png)

![image](https://cloud.githubusercontent.com/assets/4040120/10232398/2e695ea6-6891-11e5-95ed-88b62fe764a2.png)

Using the schema instance, you can generate proto:

```java
public static class TestClass {
	private int[] intArray;
	private long[] longArray;
}
```

==>

```proto
message TestClass {
  optional Int32Array intArray = 1;
  optional Int64Array longArray = 2;
}

message Int32Array {
  optional uint32 len = 1;
  repeated int32 data = 2;
}

message Int64Array {
  optional uint32 len = 1;
  repeated int64 data = 2;
}
```
